services/src/main/scala/cromwell/services/healthmonitor/HealthMonitorServiceActor.scala

```diff
@@ -0,0 +1,119 @@
+
+import akka.actor.{Actor, Scheduler, Timers}
+import akka.pattern.{after, pipe}
+import cromwell.util.GracefulShutdownHelper.ShutdownCommand
+import com.typesafe.scalalogging.LazyLogging
+
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+import scala.util.control.NonFatal
+import scala.language.postfixOps
+import HealthMonitorServiceActor._
+import com.typesafe.config.Config
+import cromwell.services.ServiceRegistryActor.ServiceRegistryMessage
+import net.ceedubs.ficus.Ficus._
+
+trait HealthMonitorServiceActor extends Actor with LazyLogging with Timers {
+  val serviceConfig: Config
+  def subsystems: List[MonitoredSubsystem]
+
+  implicit val ec: ExecutionContext = context.system.dispatcher
+
+  lazy val futureTimeout: FiniteDuration = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.check-timeout").getOrElse(DefaultFutureTimeout)
+  lazy val staleThreshold: FiniteDuration = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.status-ttl").getOrElse(DefaultStaleThreshold)
+
+  override def preStart(): Unit = {
+    val sweepTime = serviceConfig.as[Option[FiniteDuration]]("services.HealthMonitor.check-refresh-time").getOrElse(DefaultSweepTime)
+    timers.startPeriodicTimer(CheckTickKey, CheckAll, sweepTime)
+    logger.info("Starting health monitor with the following checks: " + subsystems.map(_.name).mkString(", "))
+  }
+
+  /**
+    * Contains each subsystem status along with a timestamp of when the entry was made so we know when the status
+    * goes stale. Initialized with unknown status.
+    */
+  private var data: Map[MonitoredSubsystem, CachedSubsystemStatus] = {
+    val now = System.currentTimeMillis
+    subsystems.map((_, CachedSubsystemStatus(UnknownStatus, now))).toMap
+  }
+
+  override def receive: Receive = {
+    case CheckAll => subsystems.foreach(checkSubsystem)
+    case Store(subsystem, status) => store(subsystem, status)
+    case GetCurrentStatus => sender ! getCurrentStatus
+    case ShutdownCommand => context.stop(self) // Not necessary but service registry requires it. See #2575
+  }
+
+  private def checkSubsystem(subsystem: MonitoredSubsystem): Unit = {
+    val result = subsystem.check()
+    val errMsg = s"Timed out after ${futureTimeout.toString} waiting for a response from ${subsystem.toString}"
+    result.withTimeout(futureTimeout, errMsg, context.system.scheduler)
+      .recover { case NonFatal(ex) =>
+        failedStatus(ex.getMessage)
+      } map {
+      Store(subsystem, _)
+    } pipeTo self
+
+    ()
+  }
+
+  private def store(subsystem: MonitoredSubsystem, status: SubsystemStatus): Unit = {
+    data = data + ((subsystem, CachedSubsystemStatus(status, System.currentTimeMillis)))
+    logger.debug(s"New health monitor state: $data")
+  }
+
+  private def getCurrentStatus: StatusCheckResponse = {
+    val now = System.currentTimeMillis()
+    // Convert any expired statuses to unknown
+    val processed = data map {
```

Comment:
ToL: instead of checking every time if statuses are stale, could we instead have a delayed message rescheduled after each subsystem update that would flip the status to Unknown in the store itself ?